File: Leetcode/0540 Single Element in a Sorted Array.py
class Solution:
    def singleNonDuplicate(self, a):
        lo, hi = 0, len(a) - 1
        mid = lo + (hi - lo) // 2
        while lo < hi:
            if a[mid-1] != a[mid] != a[mid+1]:
                break

            if a[mid-1] == a[mid]:
                left = (lo, mid)
                right = (mid + 1, hi)
            else:
                left = (lo, mid-1)
                right = (mid, hi)

            lo, hi = left if is_odd_sized(left) else right
            mid = lo + (hi - lo) // 2

        return a[mid]


def is_odd_sized(interval):
    return (interval[1] - interval[0] + 1) % 2 == 1

File: Leetcode/0540 Single Element in a Sorted Array.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
  public:
    int singleNonDuplicate(const vector<int> &a) {
        int lo = 0, hi = a.size() - 1;
        int mid = lo + (hi - lo) / 2;
        while (lo < hi) {
            if (a[mid - 1] != a[mid] and a[mid] != a[mid + 1]) {
                break;
            }
            pair<int, int> left, right;
            if (a[mid - 1] == a[mid]) {
                left = {lo, mid};
                right = {mid + 1, hi};
            } else {
                left = {lo, mid - 1};
                right = {mid, hi};
            }
            tie(lo, hi) = isOddSized(left) ? left : right;
            mid = lo + (hi - lo) / 2;
        }
        return a[mid];
    }

    inline bool isOddSized(const pair<int, int> &interval) {
        return (interval.second - interval.first + 1) & 1;
    }
};